plugins/@grouparoo/google-sheets/src/lib/shared/Spreadsheet.ts

```diff
@@ -0,0 +1,315 @@
+    const private_key = this.credentials.private_key
+      ?.toString()
+      .replace(/\\n/g, "\n");
+    await this.doc.useServiceAccountAuth({ client_email, private_key });
+    this.connected = true;
+  }
+
+  async load() {
+    if (this.sheet) {
+      return this.sheet;
+    }
+    await this.connect();
+    await this.loadDocInfo();
+
+    if (this.sheetId) {
+      this.sheet = this.doc.sheetsById[this.sheetId];
+    }
+    if (!this.sheet) {
+      // didn't have that id, use the first one
+      this.sheet = this.doc.sheetsByIndex[0];
+    }
+    await this.loadHeaders();
+    return this.sheet;
+  }
+
+  async loadDocInfo() {
+    try {
+      await this.doc.loadInfo(); // loads document properties and worksheets
+    } catch (error) {
+      if (error?.response?.status === 429) {
+        await this.sleep(MINUTE_IN_MILLIS);
+        await this.loadDocInfo();
+      }
+    }
+  }
+
+  async loadHeaders() {
+    try {
+      await this.sheet.loadHeaderRow();
+      return this.sheet.headerValues;
+    } catch (error) {
+      if (error?.response?.status === 429) {
+        await this.sleep(MINUTE_IN_MILLIS);
+        return await this.loadHeaders();
+      }
+    }
+  }
+
+  async read({ limit, offset }) {
+    try {
+      const results = [];
+      const sheet = await this.load();
+      if (offset < sheet.rowCount) {
+        const rows = await sheet.getRows({ limit, offset });
+        for (const row of rows) {
+          const result: any = {};
+          result["_rowNumber"] = row.rowNumber;
+          for (const header of sheet.headerValues) {
+            result[header] = row[header];
+          }
+          results.push(result);
+        }
+      }
+      return results;
+    } catch (error) {
+      if (error?.response?.status === 429) {
+        await this.sleep(MINUTE_IN_MILLIS);
+        return this.read({ limit, offset });
+      } else {
+        throw error;
+      }
+    }
+  }
+
+  async getHeaders() {
+    await this.load();
+    return this.loadHeaders();
+  }
+
+  async getAllRows() {
+    await this.load();
+    const allRows = [];
+    const headers = await this.getHeaders();
+    if (headers) {
+      let offset = 0;
+      let rows = await this.read({ limit: GOOGLE_SHEETS_ROWS_LIMIT, offset });
+      while (rows.length > 0) {
+        allRows.push(...rows);
+        offset += GOOGLE_SHEETS_ROWS_LIMIT;
+        rows = await this.read({ limit: GOOGLE_SHEETS_ROWS_LIMIT, offset });
+      }
+    }
+    return allRows;
+  }
+
+  async getRowByPrimaryKey(column, value) {
+    await this.load();
+    if (column && value) {
+      const rows = await this.getAllRows();
+      return rows.filter((row) => row[column] === value)[0] || null;
+    }
+    return null;
+  }
+
+  async getRowObjectByPrimaryKey(column, value) {
+    await this.load();
+    const row = await this.getRowByPrimaryKey(column, value);
+    if (row?._rowNumber) {
+      return await this.getRowObjectByRowNumber(row?._rowNumber);
```

Comment:
is it necessary to get the index _and then_ look up the object. This could be quite a bit of extra calls if it i further down.

Reply:
And this function is the only one calling `getRowByPrimaryKey`